File: components/backend/types/agui.go
// Package types defines AG-UI protocol types for event streaming.
// Reference: https://docs.ag-ui.com/concepts/events
package types

import (
	"encoding/json"
	"time"
)

// Timestamp helpers for AG-UI events and metadata.
//
// AG-UI protocol timestamps are epoch milliseconds (int64), not ISO strings.
// See: BaseEventSchema in @ag-ui/core — timestamp: z.number().optional()
const (
	// AGUIMetadataTimestampFormat is used for run/session metadata timestamps.
	// This is sufficient for human-readable timestamps where nanosecond precision isn't needed.
	// Format: "2006-01-02T15:04:05Z07:00" (RFC3339)
	// Used in: AGUIRunMetadata.StartedAt, AGUIRunMetadata.FinishedAt
	AGUIMetadataTimestampFormat = time.RFC3339
)

// AGUITimestampNow returns the current time as epoch milliseconds,
// which is the format expected by the AG-UI protocol.
func AGUITimestampNow() int64 {
	return time.Now().UnixMilli()
}

// AG-UI Event Types as defined in the protocol specification
// See: https://docs.ag-ui.com/concepts/events
const (
	// Lifecycle events
	EventTypeRunStarted  = "RUN_STARTED"
	EventTypeRunFinished = "RUN_FINISHED"
	EventTypeRunError    = "RUN_ERROR"

	// Step events
	EventTypeStepStarted  = "STEP_STARTED"
	EventTypeStepFinished = "STEP_FINISHED"

	// Text message events (streaming)
	EventTypeTextMessageStart   = "TEXT_MESSAGE_START"
	EventTypeTextMessageContent = "TEXT_MESSAGE_CONTENT"
	EventTypeTextMessageEnd     = "TEXT_MESSAGE_END"

	// Tool call events (streaming)
	EventTypeToolCallStart = "TOOL_CALL_START"
	EventTypeToolCallArgs  = "TOOL_CALL_ARGS"
	EventTypeToolCallEnd   = "TOOL_CALL_END"

	// State management events
	EventTypeStateSnapshot = "STATE_SNAPSHOT"
	EventTypeStateDelta    = "STATE_DELTA"

	// Message snapshot for restore/reconnect
	EventTypeMessagesSnapshot = "MESSAGES_SNAPSHOT"

	// Activity events (frontend-only durable UI)
	EventTypeActivitySnapshot = "ACTIVITY_SNAPSHOT"
	EventTypeActivityDelta    = "ACTIVITY_DELTA"

	// Raw event for pass-through
	EventTypeRaw = "RAW"

	// Custom event for platform extensions
	EventTypeCustom = "CUSTOM"

	// META event for user feedback (thumbs up/down)
	// See: https://docs.ag-ui.com/drafts/meta-events
	EventTypeMeta = "META"
)

// Agent status values derived from the AG-UI event stream.
const (
	AgentStatusWorking      = "working"
	AgentStatusIdle         = "idle"
	AgentStatusWaitingInput = "waiting_input"
)

// AG-UI Message Roles
// See: https://docs.ag-ui.com/concepts/messages
const (
	RoleUser      = "user"
	RoleAssistant = "assistant"
	RoleSystem    = "system"
	RoleTool      = "tool"
	RoleDeveloper = "developer"
	RoleActivity  = "activity"
)

// BaseEvent is the common structure for all AG-UI events
// See: https://docs.ag-ui.com/concepts/events#baseeventproperties
type BaseEvent struct {
	Type      string `json:"type"`
	ThreadID  string `json:"threadId"`
	RunID     string `json:"runId"`
	Timestamp int64  `json:"timestamp,omitempty"` // Epoch milliseconds (AG-UI spec)
	// Optional fields
	MessageID   string `json:"messageId,omitempty"`
	ParentRunID string `json:"parentRunId,omitempty"`
}

// RunAgentInput is the input format for starting an AG-UI run.
// Messages is json.RawMessage so the Go proxy passes messages through untouched
// to the Python runner. The AG-UI spec uses OpenAI-style nested tool calls
// (e.g. {id, type, function: {name, arguments}}) which would be silently lost
// if parsed into the flat Go Message/ToolCall structs.
// See: https://docs.ag-ui.com/quickstart/introduction
type RunAgentInput struct {
	ThreadID       string                 `json:"threadId,omitempty"`
	RunID          string                 `json:"runId,omitempty"`
	ParentRunID    string                 `json:"parentRunId,omitempty"`
	Messages       json.RawMessage        `json:"messages,omitempty"`
	State          map[string]interface{} `json:"state,omitempty"`
	Tools          []ToolDefinition       `json:"tools,omitempty"`
	Context        interface{}            `json:"context,omitempty"` // AG-UI sends array or object
	ForwardedProps map[string]interface{} `json:"forwardedProps,omitempty"`
}

// RunAgentOutput is the response after starting a run
type RunAgentOutput struct {
	ThreadID    string `json:"threadId"`
	RunID       string `json:"runId"`
	ParentRunID string `json:"parentRunId,omitempty"`
	StreamURL   string `json:"streamUrl,omitempty"`
}

// Message represents an AG-UI message in the conversation
// See: https://docs.ag-ui.com/concepts/messages
type Message struct {
	ID         string      `json:"id"`
	Role       string      `json:"role"`
	Content    string      `json:"content,omitempty"`
	ToolCalls  []ToolCall  `json:"toolCalls,omitempty"`
	ToolCallID string      `json:"toolCallId,omitempty"`
	Name       string      `json:"name,omitempty"`
	Timestamp  string      `json:"timestamp,omitempty"`
	Metadata   interface{} `json:"metadata,omitempty"`
}

// ToolCall represents a tool call made by the assistant
type ToolCall struct {
	ID              string `json:"id"`
	Name            string `json:"name"`
	Args            string `json:"args"`
	Type            string `json:"type,omitempty"`            // "function"
	ParentToolUseID string `json:"parentToolUseId,omitempty"` // For hierarchical nesting
	Result          string `json:"result,omitempty"`
	Status          string `json:"status,omitempty"` // "pending", "running", "completed", "error"
	Error           string `json:"error,omitempty"`
	Duration        int64  `json:"duration,omitempty"` // milliseconds
}

// ToolDefinition describes an available tool
type ToolDefinition struct {
	Name        string                 `json:"name"`
	Description string                 `json:"description,omitempty"`
	Parameters  map[string]interface{} `json:"parameters,omitempty"`
}

// RunStartedEvent is emitted when a run begins
type RunStartedEvent struct {
	BaseEvent
	Input *RunAgentInput `json:"input,omitempty"`
}

// RunFinishedEvent is emitted when a run completes successfully
type RunFinishedEvent struct {
	BaseEvent
	Output interface{} `json:"output,omitempty"`
}

// RunErrorEvent is emitted when a run fails
type RunErrorEvent struct {
	BaseEvent
	Error   string `json:"error"`
	Code    string `json:"code,omitempty"`
	Details string `json:"details,omitempty"`
}

// StepStartedEvent marks the beginning of a processing step
type StepStartedEvent struct {
	BaseEvent
	StepID   string `json:"stepId"`
	StepName string `json:"stepName"`
}

// StepFinishedEvent marks the completion of a processing step
type StepFinishedEvent struct {
	BaseEvent
	StepID   string `json:"stepId"`
	StepName string `json:"stepName"`
	Duration int64  `json:"duration,omitempty"` // milliseconds
}

// TextMessageStartEvent begins a streaming text message
type TextMessageStartEvent struct {
	BaseEvent
	Role string `json:"role"`
}

// TextMessageContentEvent contains a chunk of text content
type TextMessageContentEvent struct {
	BaseEvent
	Delta string `json:"delta"`
}

// TextMessageEndEvent marks the end of a streaming text message
type TextMessageEndEvent struct {
	BaseEvent
}

// ToolCallStartEvent begins a streaming tool call
type ToolCallStartEvent struct {
	BaseEvent
	ToolCallID      string `json:"toolCallId"`
	ToolCallName    string `json:"toolCallName"`
	ParentMessageID string `json:"parentMessageId,omitempty"`
	ParentToolUseID string `json:"parentToolUseId,omitempty"`
}

// ToolCallArgsEvent contains a chunk of tool call arguments
type ToolCallArgsEvent struct {
	BaseEvent
	ToolCallID string `json:"toolCallId"`
	Delta      string `json:"delta"`
}

// ToolCallEndEvent marks the end of a streaming tool call
type ToolCallEndEvent struct {
	BaseEvent
	ToolCallID string `json:"toolCallId"`
	Result     string `json:"result,omitempty"`
	Error      string `json:"error,omitempty"`
	Duration   int64  `json:"duration,omitempty"` // milliseconds
}

// StateSnapshotEvent provides complete state for hydration
type StateSnapshotEvent struct {
	BaseEvent
	State map[string]interface{} `json:"state"`
}

// StateDeltaEvent provides incremental state updates
type StateDeltaEvent struct {
	BaseEvent
	Delta []StatePatch `json:"delta"`
}

// StatePatch represents a JSON Patch operation for state updates
type StatePatch struct {
	Op    string      `json:"op"`   // "add", "remove", "replace"
	Path  string      `json:"path"` // JSON Pointer
	Value interface{} `json:"value,omitempty"`
}

// MessagesSnapshotEvent provides complete message history for hydration
type MessagesSnapshotEvent struct {
	BaseEvent
	Messages []Message `json:"messages"`
}

// ActivitySnapshotEvent provides complete activity UI state
type ActivitySnapshotEvent struct {
	BaseEvent
	Activities []Activity `json:"activities"`
}

// ActivityDeltaEvent provides incremental activity updates
type ActivityDeltaEvent struct {
	BaseEvent
	Delta []ActivityPatch `json:"delta"`
}

// Activity represents a durable frontend UI element
type Activity struct {
	ID       string                 `json:"id"`
	Type     string                 `json:"type"`
	Title    string                 `json:"title,omitempty"`
	Status   string                 `json:"status,omitempty"` // "pending", "running", "completed", "error"
	Progress float64                `json:"progress,omitempty"`
	Data     map[string]interface{} `json:"data,omitempty"`
}

// ActivityPatch represents an update to an activity
type ActivityPatch struct {
	Op       string   `json:"op"` // "add", "update", "remove"
	Activity Activity `json:"activity"`
}

// RawEvent allows pass-through of arbitrary data
type RawEvent struct {
	BaseEvent
	Data interface{} `json:"data"`
}

// MetaEvent represents AG-UI META events for user feedback
// See: https://docs.ag-ui.com/drafts/meta-events#user-feedback
type MetaEvent struct {
	BaseEvent
	MetaType string                 `json:"metaType"` // "thumbs_up" or "thumbs_down"
	Payload  map[string]interface{} `json:"payload"`
}

// FeedbackPayload contains the payload for feedback META events
type FeedbackPayload struct {
	MessageID string `json:"messageId,omitempty"` // ID of the message being rated
	UserID    string `json:"userId"`              // User providing feedback
	Reason    string `json:"reason,omitempty"`    // Reason for negative feedback
	Comment   string `json:"comment,omitempty"`   // Additional user comment
	// Extended fields for Langfuse context
	ProjectName       string                   `json:"projectName,omitempty"`
	SessionName       string                   `json:"sessionName,omitempty"`
	Workflow          string                   `json:"workflow,omitempty"`
	Context           string                   `json:"context,omitempty"`
	IncludeTranscript bool                     `json:"includeTranscript,omitempty"`
	Transcript        []FeedbackTranscriptItem `json:"transcript,omitempty"`
}

// FeedbackTranscriptItem represents a message in the feedback transcript
type FeedbackTranscriptItem struct {
	Role      string `json:"role"`
	Content   string `json:"content"`
	Timestamp string `json:"timestamp,omitempty"`
}

// NewBaseEvent creates a new BaseEvent with current timestamp (epoch ms).
func NewBaseEvent(eventType, threadID, runID string) BaseEvent {
	return BaseEvent{
		Type:      eventType,
		ThreadID:  threadID,
		RunID:     runID,
		Timestamp: AGUITimestampNow(),
	}
}

// WithMessageID adds a message ID to the event
func (e BaseEvent) WithMessageID(messageID string) BaseEvent {
	e.MessageID = messageID
	return e
}

// WithParentRunID adds a parent run ID to the event
func (e BaseEvent) WithParentRunID(parentRunID string) BaseEvent {
	e.ParentRunID = parentRunID
	return e
}

// AGUIEventLog represents the persisted event log structure
type AGUIEventLog struct {
	ThreadID    string      `json:"threadId"`
	RunID       string      `json:"runId"`
	ParentRunID string      `json:"parentRunId,omitempty"`
	Events      []BaseEvent `json:"events"`
	CreatedAt   string      `json:"createdAt"`
	UpdatedAt   string      `json:"updatedAt"`
}

// AGUIRunMetadata contains metadata about a run for indexing
type AGUIRunMetadata struct {
	ThreadID     string `json:"threadId"`
	RunID        string `json:"runId"`
	ParentRunID  string `json:"parentRunId,omitempty"`
	SessionName  string `json:"sessionName"`
	ProjectName  string `json:"projectName"`
	StartedAt    string `json:"startedAt"`
	FinishedAt   string `json:"finishedAt,omitempty"`
	Status       string `json:"status"` // "running", "completed", "error"
	EventCount   int    `json:"eventCount"`
	RestartCount int    `json:"restartCount,omitempty"`
}

File: components/frontend/src/types/agui.ts
/**
 * AG-UI Protocol Types
 *
 * Re-exports canonical types from @ag-ui/client (@ag-ui/core) and defines
 * platform-specific extensions for hierarchical tool calls, message metadata,
 * and streaming client state.
 *
 * Reference: https://docs.ag-ui.com/concepts/events
 * Reference: https://docs.ag-ui.com/concepts/messages
 */

// ── Core AG-UI types (re-exported from @ag-ui/client) ──

export {
  EventType,
  type BaseEvent,
  type RunAgentInput,
  type AGUIEvent,
  type Message,
  type Role,
  type ToolCall,
  type FunctionCall,
  type Tool,
  type Context,
  type State,
  type RunStartedEvent,
  type RunFinishedEvent,
  type RunErrorEvent,
  type TextMessageStartEvent,
  type TextMessageContentEvent,
  type TextMessageEndEvent,
  type ToolCallStartEvent,
  type ToolCallArgsEvent,
  type ToolCallEndEvent,
  type ToolCallResultEvent,
  type StateSnapshotEvent,
  type StateDeltaEvent,
  type MessagesSnapshotEvent,
  type StepStartedEvent,
  type StepFinishedEvent,
  type ReasoningStartEvent,
  type ReasoningMessageStartEvent,
  type ReasoningMessageContentEvent,
  type ReasoningMessageEndEvent,
  type ReasoningEndEvent,
  type CustomEvent as AGUICustomEvent,
} from '@ag-ui/client'

import { EventType } from '@ag-ui/client'
import type {
  ToolCall,
  Message,
  RunStartedEvent,
  RunFinishedEvent,
  RunErrorEvent,
  TextMessageStartEvent,
  TextMessageContentEvent,
  TextMessageEndEvent,
  ToolCallStartEvent,
  ToolCallEndEvent,
  StateSnapshotEvent,
  MessagesSnapshotEvent,
} from '@ag-ui/client'

// ── Platform Extension: PlatformToolCall ──
// Extends core ToolCall with platform-specific tracking fields
// for hierarchical tool calls (sub-agents), result caching, and timing.
export type PlatformToolCall = ToolCall & {
  parentToolUseId?: string
  result?: string
  status?: 'pending' | 'running' | 'completed' | 'error'
  error?: string
  duration?: number
}

// ── Platform Extension: PlatformMessage ──
// Extends core Message union with platform-specific fields.
// Because Message is a discriminated union (A | B | C), the intersection
// distributes: (A & Ext) | (B & Ext) | (C & Ext), preserving discrimination.
/** Structured content block for reasoning messages */
export type ReasoningContent = {
  type: 'reasoning_block'
  thinking: string
  signature: string
}

/** Message metadata for sender attribution in multi-user sessions */
export type MessageMetadata = {
  senderId?: string
  senderDisplayName?: string
  hidden?: boolean
  [key: string]: unknown
}

export type PlatformMessage = Message & {
  timestamp?: string
  metadata?: MessageMetadata
  content?: string | ReasoningContent  // Widened to support structured reasoning blocks
  name?: string  // Tool name (not on core ToolMessage, but platform sends it)
  toolCalls?: PlatformToolCall[]
  toolCallId?: string  // Present on tool-role messages
  parentToolUseId?: string
  children?: PlatformMessage[]
}

// ── Wire event types for platform-specific fields ──
// The backend sends extra fields not in the core schema. Since AG-UI schemas
// use "passthrough", these survive JSON parsing but need explicit types.

/** ToolCallStartEvent with backend's extra snake_case parent field */
export type WireToolCallStartEvent = ToolCallStartEvent & {
  parent_tool_call_id?: string
}

/** Platform raw event — extends core RawEvent with optional data field.
 *  The backend may send payload in either `event` or `data`. */
export type PlatformRawEvent = {
  type: typeof EventType.RAW
  timestamp?: number
  event?: Record<string, unknown>
  data?: Record<string, unknown>
  source?: string
}

// ── Platform Activity types ──
// The core ActivitySnapshotEvent/ActivityDeltaEvent are per-message, not
// array-based. The platform uses an array-based model for UI rendering.

export type PlatformActivity = {
  id: string
  type: string
  title?: string
  status?: 'pending' | 'running' | 'completed' | 'error'
  progress?: number
  data?: Record<string, unknown>
}

export type PlatformActivityPatch = {
  op: 'add' | 'update' | 'remove'
  activity: PlatformActivity
}

/** Platform's array-based ACTIVITY_SNAPSHOT (differs from core's per-message model) */
export type PlatformActivitySnapshotEvent = {
  type: typeof EventType.ACTIVITY_SNAPSHOT
  timestamp?: number
  activities?: PlatformActivity[]
}

/** Platform's array-based ACTIVITY_DELTA (differs from core's per-message model) */
export type PlatformActivityDeltaEvent = {
  type: typeof EventType.ACTIVITY_DELTA
  timestamp?: number
  delta: PlatformActivityPatch[]
}

// ── Platform-specific types (no core equivalent) ──

// Meta event (user feedback, annotations, etc.)
export type AGUIMetaEvent = {
  type: 'META'
  metaType: string
  payload: Record<string, unknown>
  threadId: string
  ts?: number
}

// Union of all events the platform handles (core + META)
export type PlatformEvent = import('@ag-ui/client').AGUIEvent | AGUIMetaEvent

// Pending tool call during streaming (flat format for accumulation)
export type PendingToolCall = {
  id: string
  name: string
  args: string
  parentToolUseId?: string
  parentMessageId?: string
  timestamp?: string
}

// Feedback type for messages
export type MessageFeedback = 'thumbs_up' | 'thumbs_down'

// Client state for AG-UI streaming
export type AGUIClientState = {
  threadId: string | null
  runId: string | null
  status: 'idle' | 'connecting' | 'connected' | 'error' | 'completed'
  messages: PlatformMessage[]
  state: Record<string, unknown>
  activities: PlatformActivity[]
  currentMessage: {
    id: string | null
    role: string | null
    content: string
    timestamp?: string
  } | null
  // DEPRECATED: Use pendingToolCalls instead for parallel tool call support
  currentToolCall: {
    id: string | null
    name: string | null
    args: string
    parentToolUseId?: string
  } | null
  // Track ALL in-progress tool calls (supports parallel tool execution)
  pendingToolCalls: Map<string, PendingToolCall>
  // Track child tools that finished before their parent
  pendingChildren: Map<string, PlatformMessage[]>
  error: string | null
  // Track feedback for messages (messageId -> feedback type)
  messageFeedback: Map<string, MessageFeedback>
  // Buffer for reasoning content during streaming
  currentReasoning: {
    id: string | null
    content: string
    timestamp?: string
  } | null
  // Buffer for thinking content during streaming (THINKING_* events)
  currentThinking: {
    id: string | null
    content: string
    timestamp?: string
  } | null
  // Background tasks tracked from CustomEvent task:* events
  backgroundTasks: Map<string, import('@/types/background-task').BackgroundTask>
}

// ── Type Guards ──
// Narrow parsed SSE events to specific core event types.

export function isRunStartedEvent(event: { type: string }): event is RunStartedEvent {
  return event.type === EventType.RUN_STARTED
}

export function isRunFinishedEvent(event: { type: string }): event is RunFinishedEvent {
  return event.type === EventType.RUN_FINISHED
}

export function isRunErrorEvent(event: { type: string }): event is RunErrorEvent {
  return event.type === EventType.RUN_ERROR
}

export function isTextMessageStartEvent(event: { type: string }): event is TextMessageStartEvent {
  return event.type === EventType.TEXT_MESSAGE_START
}

export function isTextMessageContentEvent(event: { type: string }): event is TextMessageContentEvent {
  return event.type === EventType.TEXT_MESSAGE_CONTENT
}

export function isTextMessageEndEvent(event: { type: string }): event is TextMessageEndEvent {
  return event.type === EventType.TEXT_MESSAGE_END
}

export function isToolCallStartEvent(event: { type: string }): event is ToolCallStartEvent {
  return event.type === EventType.TOOL_CALL_START
}

export function isToolCallEndEvent(event: { type: string }): event is ToolCallEndEvent {
  return event.type === EventType.TOOL_CALL_END
}

export function isStateSnapshotEvent(event: { type: string }): event is StateSnapshotEvent {
  return event.type === EventType.STATE_SNAPSHOT
}

export function isMessagesSnapshotEvent(event: { type: string }): event is MessagesSnapshotEvent {
  return event.type === EventType.MESSAGES_SNAPSHOT
}

export function isActivitySnapshotEvent(event: { type: string }): event is PlatformActivitySnapshotEvent {
  return event.type === EventType.ACTIVITY_SNAPSHOT
}
